src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceManager)), Shared]
+internal class DiagnosticSourceManager : IDiagnosticSourceManager
+{
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _documentProviders;
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _workspaceProviders;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<string> GetSourceNames(bool isDocument)
+        => (isDocument ? _documentProviders : _workspaceProviders).Keys;
+
+    /// <inheritdoc />
+    public async ValueTask<ImmutableArray<IDiagnosticSource>> CreateDiagnosticSourcesAsync(RequestContext context, string? sourceName, bool isDocument, CancellationToken cancellationToken)
+    {
+        var providersDictionary = isDocument ? _documentProviders : _workspaceProviders;
+        if (sourceName != null)
+        {
+            Contract.ThrowIfFalse(providersDictionary.TryGetValue(sourceName, out var provider), $"Unrecognized source {sourceName}");
+            return await provider.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            // VS Code (and legacy VS ?) pass null sourceName when requesting all sources.
+            using var _ = ArrayBuilder<IDiagnosticSource>.GetInstance(out var sourcesBuilder);
+            foreach (var kvp in providersDictionary)
+            {
+                // Exclude task diagnostics from the aggregated sources.
+                if (kvp.Key != PullDiagnosticCategories.Task)
+                {
+                    var namedSources = await kvp.Value.CreateDiagnosticSourcesAsync(context, cancellationToken).ConfigureAwait(false);
+                    sourcesBuilder.AddRange(namedSources);
+                }
+            }
+
+            var sources = sourcesBuilder.ToImmutableAndClear();
+            if (!isDocument || sources.Length <= 1)
```

Comment:
for workspace sources we still need to aggregate them *per* document.  We can't report diagnostics for the same document twice for ws diagnostics (the later report will overwrite the first)